pkg/querier/remote_read_test.go

```diff
@@ -208,280 +298,514 @@ func TestRemoteReadHandler_Samples(t *testing.T) {
-
-					i := 0
-					for {
-						var res prompb.ChunkedReadResponse
-						err := stream.NextProto(&res)
-						if errors.Is(err, io.EOF) {
-							break
-						}
-						require.NoError(t, err)
+			requestBody, err := proto.Marshal(&prompb.ReadRequest{
+				Queries:               testData.query,
+				AcceptedResponseTypes: []prompb.ReadRequest_ResponseType{prompb.ReadRequest_STREAMED_XOR_CHUNKS},
+			})
+			require.NoError(t, err)
+			requestBody = snappy.Encode(nil, requestBody)
+			request, err := http.NewRequest(http.MethodPost, "/api/v1/read", bytes.NewReader(requestBody))
+			require.NoError(t, err)
+			request.Header.Set("X-Prometheus-Remote-Read-Version", "0.1.0")
 
-						if len(testData.expectedResults) < i+1 {
-							require.Fail(t, "unexpected result message")
-						}
-						require.Equal(t, testData.expectedResults[i], &res)
-						i++
-					}
-					require.Len(t, testData.expectedResults, i)
-
-					// Ensure the time range passed down to the queryable is the expected one.
-					require.Equal(t, queryData.expectedQueriedStart, actualQueriedStart)
-					require.Equal(t, queryData.expectedQueriedEnd, actualQueriedEnd)
+			recorder := httptest.NewRecorder()
+			handler.ServeHTTP(recorder, request)
+
+			require.Equal(t, 200, recorder.Result().StatusCode)
+			require.Equal(t, []string{api.ContentTypeRemoteReadStreamedChunks}, recorder.Result().Header["Content-Type"])
+
+			stream := prom_remote.NewChunkedReader(recorder.Result().Body, config.DefaultChunkedReadLimit, nil)
+
+			var responses []*prompb.ChunkedReadResponse
+			for {
+				var res prompb.ChunkedReadResponse
+				err := stream.NextProto(&res)
+				if errors.Is(err, io.EOF) {
+					break
+				}
+				require.NoError(t, err)
+				responses = append(responses, &res)
+			}
+
+			actualResponsesByQuery := map[int64][]*prompb.ChunkedReadResponse{}
+			expectedResponsesByQuery := map[int64][]*prompb.ChunkedReadResponse{}
+			for _, expectedResponse := range testData.expectedResults {
+				for _, response := range expectedResponse.responses {
+					expectedResponsesByQuery[response.QueryIndex] = append(expectedResponsesByQuery[response.QueryIndex], response)
+				}
+			}
+			for _, actualResponse := range responses {
+				actualResponsesByQuery[actualResponse.QueryIndex] = append(actualResponsesByQuery[actualResponse.QueryIndex], actualResponse)
+			}
+
+			require.Len(t, expectedResponsesByQuery, len(actualResponsesByQuery))
+			for queryID, expectedResponses := range expectedResponsesByQuery {
+				require.Equal(t, expectedResponses, actualResponsesByQuery[queryID])
+			}
+
+			// Use elements matching to verify the expected time ranges
+			var expectedTimeRanges []queryStartEnd
+			for _, expected := range testData.expectedResults {
+				expectedTimeRanges = append(expectedTimeRanges, queryStartEnd{
+					start: expected.start,
+					end:   expected.end,
 				})
 			}
+			require.ElementsMatch(t, expectedTimeRanges, actualQueryTimeRanges)
```

Comment:
Why `ElementsMatch` and not `Equal` here? Don't we care about the order?

Reply:
i couldn't find anything in the remote read spec to indicate an order of the results. Each result has a QueryIndex field, so clients can differentiate between the two queries.